data_steward/cdr_cleaner/cleaning_rules/null_invalid_foreign_keys.py

```diff
@@ -21,84 +29,215 @@
+    def get_affected_tables(self):
+        """
+        This method gets all the tables that are affected by this cleaning rule
+        
+        :return: list of affected tables
+        """
+        return resources.CDM_TABLES
+
+    def get_mapping_table(self, domain_table):
+        """
+        Get name of mapping table generated for a domain table
+
+        :param domain_table: one of the domain tables (e.g. 'visit_occurrence',
+            'condition_occurrence')
+        :return: mapping table name
+        """
+        return '_mapping_' + domain_table
+
+    def get_field_names(self, table):
+        """
+        This method gets the list of field names in a table affected by the cleaning rule
+
+        :param table: single table in the list of affected tables
+        :return: list of field names in a single affected table
+        """
         field_names = [field['name'] for field in resources.fields_for(table)]
-        foreign_keys_flags = []
-        fields_to_join = []
+        return field_names
+
+    def get_foreign_keys(self, table):
+        """
+        This method gets the list of foreign keys determined from the list of field names
 
-        for field_name in field_names:
+        :param table: single table in the list of affected tables
+        :return: list of foreign keys
+        """
+        foreign_keys_flags = []
+        for field_name in self.get_field_names(table):
             if field_name in FOREIGN_KEYS_FIELDS and field_name != table + '_id':
-                fields_to_join.append(field_name)
                 foreign_keys_flags.append(field_name)
+        return foreign_keys_flags
+
+    def has_foreign_key(self, table):
+        """
+        This method determines if a table contains has a foreign key
 
-        if fields_to_join:
-            col_exprs = []
-            for field in field_names:
-                if field in fields_to_join:
-                    if field in foreign_keys_flags:
-                        col_expr = '{x}.'.format(x=field[:3]) + field
+        :param table: single table in the list of affected tables
+        :return: true, if the table contains a foreign key, false if the table does not contain a foreign key
+        """
+        return len(self.get_foreign_keys(table)) > 0
+
+    def get_col_expression(self, table):
+        """
+        This method formats the column name depending of if it is a foreign key or not. If the column is a foreign key
+        it will be prefixed with the first part of the table name. If the column is not a foreign key, there will be
+        no changes to the column name.
+
+        :param table: single table in the list of affected tables
+        :return: all the formatted column names for the table
+        """
+        col_exprs = []
+        foreign_keys = self.get_foreign_keys(table)
+        for field in self.get_field_names(table):
+            if field in foreign_keys:
+                col_expr = '{x}.'.format(x=field[:3]) + field
```

Comment:
```suggestion
                col_expr = f'{field[:3]}.field'
```

Reply:
This isn't working as listed, changed to this, `col_expr = f'{field[:3]}.{field}'`, so that `field` is a variable, not printed string. This was resulting in the `col_expr` to print columns as 'loc.field' instead of 'loc.location_id'